File: bin/golden_test.pl
#!/usr/bin/perl
use strict;

# Compares FFT log file from rtl simulation, against a golden model.
# Summarizes results from each and says either PASS or FAIL.

# This script should produce same results as the alternative python version
# "golden_test.py". E.g. to compare the two
# 
#   f=../tst/regress/old/verlint/fft.log
#   golden_test.pl $f 8 1 2port > tmp.pl
#   golden_test.py $f 8 1 2port > tmp.py
#   diff tmp.p[ly]

# Set one or more of these vars to "1" to test the script
my $DBG = 0;
my $DBG9 = 0;

# Help message
if (@ARGV != 4) {
    my $errmsg = sprintf(
        "\n".
        "Generates correct fft answers and compares vs. <logfile>\n".
        "\n".
        "Usage:\n".
        "   $0 <logfile> <npoints> <n_butterfly_units> <sram_type>\n".
        "\n".
        "Examples:\n".
        "    $0 fft.log  8    1 \"2port\"\n".
        "    $0 simv.log 1024 4 \"1port\"\n".
        "\n".
        "To test failure mechanism, use sram_type \"test_error\"\n".
        "    $0 fft.log  8    1 \"test_error\"\n".
        "\n".
        "    $0 ../tst/regress/old/verlint/fft.log 8 1 \"2port\"\n".
        "\n");
    die $errmsg;
}

# Process command-line arguments
my ($LOGFILE, $npoints,$nunits,$sram_type) = ($ARGV[0], $ARGV[1], $ARGV[2], $ARGV[3]);

print "Comparing FFT log '$LOGFILE' vs. gold model (GM)\n";
print "with npoints=$npoints, nunits=$nunits, sram_type='$sram_type'\n";


########################################################################
# Set FFTGEN_DIR as the path to the local fftgen repo.
# Assumes that this script lives in FFTGEN_DIR/bin
# E.g. scriptdir='/home/steveri/github/fftgen/bin'
my $FFTGEN_DIR = mydir("..");
my $RUN_DIR = ".";


# Compare results in "$FFTGEN_DIR/$LOGFILE" against result of running golden model
do_test($npoints, $nunits, $sram_type);
exit(0);

sub do_test {
    my $npoints = shift;
    my $nunits = shift;
    my $sram_type = shift;
    my $DBG = 0;

    # Compare results in <LOGFILE> against result of running golden model

    # Examples:
    #   do_test  8, 1, "2port"
    #   do_test 16, 4, "1port"

    if ($DBG) { print('do_test %s %s %s\n', $npoints, $nunits, $sram_type); }

    # Run the golden model and fetch the results
    my @gm_results = golden_model($npoints, $nunits);
    if ($DBG) {
        print 'gm_results='; print(@gm_results); print "\n"; 
        foreach my $result (@gm_results) { print "bat $result\n"; }
        print "\n";
    }

    my @fft_results = do_fft();  # Fetch results found in fft log <LOGFILE>
    print "\n";

    if ($DBG) {
        foreach my $result (@fft_results) { print "fat $result\n\n"; } print "\n";
    }

    # If lengths don't match, maybe user got command line args wrong (???)
    # gm_results=(0,1,2,3,4,5,6,7,8); # Uncomment to check error message
    if ($#gm_results != $#fft_results) {

        my $errmsg = sprintf("ERROR Golden model produced %d results and FFT produced %d.\n",
                            ($#gm_results+1)/2, ($#fft_results+1)/2);
        $errmsg .= "Maybe try this:\n";
        $errmsg .= "$FFTGEN_DIR/bin/fft_golden_model.pl $npoints $nunits\n";
        die $errmsg;
        exit(-1);
    }

    # Secret sram type to test error handling; inserts a single faulty datapoint
    if ($sram_type eq "test_error") {
        $sram_type = "1port";
        $fft_results[3] = "99999";   # Guarantees at least one error
        # $gm_results[3] = "42";
    }

    my $nerrors = 0;
    $nerrors += print_results("Real", $npoints, \@gm_results, \@fft_results);
    $nerrors += print_results("Imag", $npoints, \@gm_results, \@fft_results);

    if ($DBG) { print('found', $nerrors, 'errors\n'); }

    my $test_result = ($nerrors==0) ? "PASS" : "FAIL";
    my $plural = ($nerrors==1) ? "" : "s";
    print "$test_result: $nerrors mismatched result$plural --- $npoints $nunits $sram_type\n";
}

sub do_fft {

    ##############################################################################
    # Print results from FFT log file

    my $fft = `cat $RUN_DIR/$LOGFILE | awk -f $FFTGEN_DIR/bin/process_test5.awk | egrep '^ix'`;

    # Above command should have yielded e.g.
    #
    # ix00: SRAM000 t5 13500: Wrote(f) wr_data_i mem[0] <= ( 8.000000, 0.000000)
    # ix01: SRAM003 t5 13500: Wrote(f) wr_data_i mem[0] <= ( 1.000000,-5.027340)
    # ix02: SRAM006 t5 14500: Wrote(f) wr_data_i mem[0] <= ( 0.000000, 0.000000)
    # ix03: SRAM005 t5 14500: Wrote(f) wr_data_i mem[0] <= ( 1.000000,-1.496606)
    # ix04: SRAM004 t5 15500: Wrote(f) wr_data_i mem[0] <= ( 0.000000, 0.000000)
    # ix05: SRAM007 t5 15500: Wrote(f) wr_data_i mem[0] <= ( 1.000000,-0.668179)
    # ix06: SRAM002 t5 16500: Wrote(f) wr_data_i mem[0] <= ( 0.000000, 0.000000)
    # ix07: SRAM001 t5 16500: Wrote(f) wr_data_i mem[0] <= ( 1.000000,-0.198912)

    # Optional debugging
    if ($DBG) { print($fft); }

    # Build ar, ai arrays containing real and imaginary numbers in above lines

    my @ar = (); # Real part of complex number
    my @ai = (); # Imaginary part of complex number
    foreach my $line (split("\n", $fft)) {

        # From sample line
        # "ix05: SRAM007 t5 15500: Wrote(f) wr_data_i mem[0] <= ( 1.000000,-0.668179)"
        # Want to extract (ix,real,imag) = ("5","1.000000","-0.668179")
 
        if ($line =~ /^ix([0-9]+).*\( *([^ ,]*), *([^ ,]*).$/) {
            my ($ix,$real,$imag) = ($1,$2,$3);
            #print "bar    $ix $real $imag\n";
            $ar[$ix] = $real;
            $ai[$ix] = $imag;
        }
    }
    if ($DBG) {
        print('real\n'); print @ar;
        print('imag\n'); print @ai;
        print "      _____________FFT____________\n";

        #    my $i = 0;
        #    my @ar = ();
        #    my @ai = ();
        #    foreach my $bank (sort (keys %ar)) {
        for (my $ix = 0; $ix < @ar; $ix++) {
            #print "baz \"$ix\" r=$ar{$ix} i=$ai{$ix}\n";
            printf("    %4d   %9s    %9s\n", $ix, $ar[$ix], $ai[$ix]);
            #printf("    %4d   %6.3f    %6.3f\n", $i, $ar[$i], $ai[$i]);
            #        @ar = (@ar, $ar{$bank});
            #        @ai = (@ai, $ai{$bank});
        }
    }
    return (@ar,@ai);
}

sub golden_model {
    my $npoints = shift;
    my $nunits = shift;

    # Low-level debugging can muck things up here.
    if ($ENV{ALL_LDBG}) {
        print STDERR "WARNING\n";
        print STDERR "WARNING Env var ALL_LDBG is set (=$ENV{ALL_LDBG})".
            ", that can screw up the golden model.\n";
        print STDERR "WARNING\n\n";
    }

    # Invoke the Golden Model and capture the results

    # E.g. my $gm = `$golden_model 8 1`;
    # Add egrep filter to eschew any lines with non-numeric
    # characters (probably not strictly necessary).
    my $golden_model = "$FFTGEN_DIR/bin/fft_golden_model.pl";
    my $gm = `$golden_model $npoints $nunits | egrep -v '[a-zZ-Z]'`;
    if ($DBG) {
        print "\n";
        print "     ______GOLDEN MODEL______\n";
        print $gm;
        print "\n";
    }
    # Example:
    #      ______GOLDEN MODEL______
    #        0    4.000     0.000
    #        1    1.000    -2.414
    #        2    0.000     0.000
    #        3    1.000    -0.414
    #        4    0.000     0.000
    #        5    1.000     0.414
    #        6    0.000     0.000
    #        7    1.000     2.414

    # Build ar, ai arrays containing real and imaginary numbers from each line.
    # Each line looks like this: " 1   1.000   -2.414"

    my @ar = ();
    my @ai = ();
    foreach my $line (split("\n", $gm)) {
        if ($line =~ /\s*(\S+)\s*(\S+)\s*(\S+)/) {
            my ($ix,$real,$imag) = ($1,$2,$3);
            #print "foo found $real and $imag\n";
            @ar = (@ar,$real);
            @ai = (@ai,$imag);
        }
    }
    if (($#ar+1) != $npoints) {
        die "ERROR Golden model produced $#ar results but npoints=$npoints\n";
        exit(-1);
    }
    return (@ar,@ai);
}

sub print_results {
    my $rtype = shift; # "Real" or "Imag"
    my $npoints = shift; 
    my $gm_results = shift;  # Pointer to results e.g. \@gm_results
    my $fft_results = shift; # Pointer to results e.g. \@fft_results

    # Compare GM/FFT real or imaginary results according to rtype "Real" or not

    my $nerrors = 0;
    my $nwarnings = 0;

    # E.g. if npoints=8, then [0..7] are real nums and [8,15] are corresponding imag pts.
    my ($first,$last) = ($rtype eq "Real") ? (0, $npoints) : ($npoints, 2*$npoints);

    print "$rtype results:      GM          FFT\n";

    # Each result MUST match to at least one decimal place;
    # emit warning if errors at two or three d.p.

    for (my $i=$first; $i < $last; $i++) {
        #printf("             %6s    %9s\n", $gm_results[$i], $fft_results[$i]);
        #printf("             %6.3f    %6.3f\n", $gm_results[$i]+0, $fft_results[$i]+0);

        my $gr =  @{$gm_results}[$i]+0;
        my $fr = @{$fft_results}[$i]+0;
    
        # Enable this block to test the warnings and errors when npoints=8 and nunits=1
        # Correct answer is i[11,13,15] = -2.414, 0.404, 2.414
        if (0) {
            if ($i == 11) { $fr = '   -0.415'; } # 3-fig warn vs. -2.414
            if ($i == 13) { $fr = '    0.424'; } # 2-fig warn vs.  0.414
            if ($i == 15) { $fr = '    2.511'; } # ERROR vs. 2.414
        }

        # Results to three sig figs; mismatch = WARNING
        my $gr3 = sprintf("%9.3f", $gr);
        my $fr3 = sprintf("%9.3f", $fr);

        # Oh this is annoying; must canonicalize "-0.000" => "0.000"
        $gr3 =~ s/[-]0[.]000/ 0.000/;
        $fr3 =~ s/[-]0[.]000/ 0.000/;

        # Results to two sig figs; mismatch = WARNING
        my $gr2 = sprintf("%9.2f", $gr3);
        my $fr2 = sprintf("%9.2f", $fr3);

        # Results to one sig figs; mismatch = ERROR
        my $gr1 = sprintf("%9.1f", $gr2);
        my $fr1 = sprintf("%9.1f", $fr2);

        # Should match to three decimal places; two or one means WARNING
        # Warning if no match @ three sig figs.
        my $status = "";

        # Confidence?
        if ($gr1 ne $fr1) {
            $status = " *** ERROR '$gr1' ne '$fr1'";
            $nerrors++;
        }
        elsif ($gr2 ne $fr2) {
            $status = " *** WARNING mismatch at 2nd sig fig";
            $nwarnings++;
        }
        elsif ($gr3 ne $fr3) {
            $status = " *** WARNING mismatch at 3rd sig fig";
            $nwarnings++;
        }
        printf("             %s    %s$status\n", $gr3, $fr3);
    }
    print "\n";
    return $nerrors
} # sub print_results

sub mydir {
    my $offset = shift;  # E.g. "../rtl"

    # Return full path of directory where this script lives, offset according to $offset.
    use Cwd 'abs_path';
    my $fullpath = abs_path($0); # Full pathname of script e.g. "/foo/bar/opendesign.pl"

    use File::Basename 'fileparse';
    my ($filename, $dir, $suffix) = fileparse($fullpath);
    return abs_path("$dir/$offset");
}

File: bin/golden_test.py
#!/usr/bin/python3
import os
import sys
import re

# Compares FFT log file from rtl simulation, against a golden model.
# Summarizes results from each and says either PASS or FAIL.

# This script should produce same results as the prior perl version
# "golden_test.pl". E.g. to compare the two
# 
#   f=../tst/regress/old/verlint/fft.log
#   golden_test.pl $f 8 1 2port > tmp.pl
#   golden_test.py $f 8 1 2port > tmp.py
#   diff tmp.p[ly]

# Set one or more of these vars to "1" to test the script
DBG = 0
DBG9 = 0

# Help message
if (len(sys.argv) != 5):
    scriptname = sys.argv[0];
    print(f"""
    DESCRIPTION: Generates correct fft answers and compares vs. <logfile>

    USAGE:
        {scriptname} <logfile> <npoints> <n_butterfly_units> <sram_type>

    Examples:
      {scriptname} fft.log  8    1 2port
      {scriptname} simv.log 1024 4 1port
      {scriptname} ../tst/regress/old/verlint/fft.log 8 1 2port

    To test failure mechanism, use sram_type "test_error"
      {scriptname} fft.log  8    1 test_error
      

    """)
    exit(13)

# Process command-line arguments
[ LOGFILE, npoints, nunits, sram_type ] = sys.argv[1:]

print(f'''
Comparing FFT log '{LOGFILE}' vs. gold model (GM)
with npoints={npoints}, nunits={nunits}, sram_type='{sram_type}'\
''')

########################################################################
# Set FFTGEN_DIR as the path to the local fftgen repo.
# Assumes that this script lives in FFTGEN_DIR/bin
# E.g. scriptdir='/home/steveri/github/fftgen/bin'
scriptdir=os.path.dirname(os.path.realpath(__file__))
FFTGEN_DIR=os.path.dirname(scriptdir)

# Compare results in "$FFTGEN_DIR/$LOGFILE" against result of running golden model

def main():
    do_test(npoints, nunits, sram_type, DBG=0); exit()

def do_test(npoints, nunits, sram_type, DBG=0):
    '''Compare results in <LOGFILE> against result of running golden model'''

    # Examples:
    #   do_test  8, 1, "2port"
    #   do_test 16, 4, "1port"

    if DBG: print('do_test', npoints, nunits, sram_type)

    # Run the golden model and fetch the results
    gm_results = golden_model(npoints, nunits)
    if DBG: print('gm_results=', gm_results)

    fft_results = do_fft()      # Fetch results found in fft log <LOGFILE>
    print('')  # ? yeah i dunno

    if DBG: print('fft_results=', fft_results)

    # If lengths don't match, maybe user got command line args wrong (???)
    # gm_results=range(9) # Uncomment to check error message
    quick_check(LOGFILE, len(gm_results), len(fft_results))

    # Secret sram type to test error handling; inserts a single faulty datapoint
    if (sram_type == "test_error"):
        sram_type = "1port"
        fft_results[3] = "99999"   # Guarantees at least one error

    nerrors = 0
    nerrors += print_results("Real", npoints, gm_results, fft_results)
    nerrors += print_results("Imag", npoints, gm_results, fft_results)

    if DBG: print('found', nerrors, 'errors')

    test_result = "PASS" if (nerrors == 0) else "FAIL"
    plural      = "s"    if (nerrors != 1) else ""
    print(f"{test_result}: {nerrors:d} mismatched result{plural} --- {npoints} {nunits} {sram_type}")

def quick_check(LOGFILE, ngr, nfr):
    'Check to see that gm and fft got same number of results e.g. 16 if npoints=8'
    # If lengths don't match, maybe user got command line args wrong (???)
    # gm_results=range(9) # Uncomment to check error message
    if (ngr != nfr):
        print(f'''\
ERROR: Golden model produced {ngr} results and FFT produced {nfr}.
   Double-check to see that number of results in '{LOGFILE}'
   matches your command-line specification 'npoints={npoints}'.
''')
        exit(13)

def do_fft(DBG=0):

    ##############################################################################
    # Print results from FFT log file

    cmd = f"cat ./{LOGFILE} | awk -f {FFTGEN_DIR}/bin/process_test5.awk | egrep '^ix'"
    fft = my_syscall(cmd)

    # Above command should have yielded fft = e.g.
    #
    # ix00: SRAM000 t5 13500: Wrote(f) wr_data_i mem[0] <= ( 8.000000, 0.000000)
    # ix01: SRAM003 t5 13500: Wrote(f) wr_data_i mem[0] <= ( 1.000000,-5.027340)
    # ix02: SRAM006 t5 14500: Wrote(f) wr_data_i mem[0] <= ( 0.000000, 0.000000)
    # ix03: SRAM005 t5 14500: Wrote(f) wr_data_i mem[0] <= ( 1.000000,-1.496606)
    # ix04: SRAM004 t5 15500: Wrote(f) wr_data_i mem[0] <= ( 0.000000, 0.000000)
    # ix05: SRAM007 t5 15500: Wrote(f) wr_data_i mem[0] <= ( 1.000000,-0.668179)
    # ix06: SRAM002 t5 16500: Wrote(f) wr_data_i mem[0] <= ( 0.000000, 0.000000)
    # ix07: SRAM001 t5 16500: Wrote(f) wr_data_i mem[0] <= ( 1.000000,-0.198912)

    # Optional debugging
    if DBG: print(cmd)
    if DBG: print(fft)

    # Build ar, ai arrays containing real and imaginary numbers in above lines

    ar = []   # Real parts of complex numbers
    ai = []   # Imaginary parts of complex numbers
    for line in fft.split('\n'):

        # From sample line
        # "ix05: SRAM007 t5 15500: Wrote(f) wr_data_i mem[0] <= ( 1.000000,-0.668179)"
        # Want to extract (ix,real,imag) = ("5","1.000000","-0.668179")

        parse = re.search(r'^ix([0-9]+).*\( *([^ ,]*), *([^ ,]*).$', line)
        if parse:
            [ ix, real, imag ] = [ parse.group(1), parse.group(2), parse.group(3) ]

            # E.g. if index ix=3, we should be about to push the fourth number in each array
            assert len(ar) == int(ix); ar = ar + [ real ]
            assert len(ai) == int(ix); ai = ai + [ imag ]

    # Debugging
    if DBG:
        print('real', ar)
        print('imag', ai)
        print("     ______________FFT___________")
        for ix in range(0,len(ar)):
            print(f"    {ix:4d} {ar[ix]:>9s}    {ai[ix]:>9s}")

    return (ar + ai)

def golden_model(npoints, nunits, DBG=0):
    '''Can use either perl or python golden model!'''
  # return golden_model_perl( npoints, nunits, DBG=0)
    return golden_model_numpy(npoints, nunits, DBG=0)

def golden_model_numpy(npoints, nunits, DBG=0):     # Note: 'nunits' is unused!
    '''Calculate FFT of a step-down function'''

    # Build a step-function test array of input datapoints, e.g. for npoints=8:
    # test_array = [1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0]
    n2 = int(npoints) >> 1  # Divide by two and still be integer
    test_array = ([1.0] * n2) + ([0.0] * n2)
    if DBG:
        print("")
        print("test_array = ", end='')
        pprint(test_array)
        print("")

    # Calculate the fft
    from numpy import fft; fft_results = fft.fft(test_array)

    # Unpack results into separate real and imaginary arrays
    ar=[]; ai=[]
    for c in fft_results:
        ar = ar + [ c.real ]
        ai = ai + [ c.imag ]

    # Optional debugging
    if DBG:
        print("")
        print("     ______GOLDEN MODEL______")
        for i in range(npoints):
            print(f"  {i:4d} {ar[i]:9.3f} {ai[i]:9.3f}")
        print("")

    return (ar+ai)

def golden_model_perl(npoints, nunits, DBG=0):
    '''Invoke the Golden Model and capture the results'''

    # E.g. my gm = my_syscall(f"{golden_model} 8 1")
    # Add egrep filter to eschew any lines with non-numeric
    # characters (probably not strictly necessary).
    golden_model = f"{FFTGEN_DIR}/bin/fft_golden_model.pl"
    cmd = f"{golden_model} {npoints} {nunits} | egrep -v '[a-zZ-Z]'"
    gm = my_syscall(cmd)
    if DBG:
        print("")
        print("     ______GOLDEN MODEL______")
        print(gm)
        print("")

    # Example:
    #      ______GOLDEN MODEL______
    #        0    4.000     0.000
    #        1    1.000    -2.414
    #        2    0.000     0.000
    #        3    1.000    -0.414
    #        4    0.000     0.000
    #        5    1.000     0.414
    #        6    0.000     0.000
    #        7    1.000     2.414

    # Build ar, ai arrays containing real and imaginary numbers from each line.
    # Each line looks like this: " 1   1.000   -2.414"

    ar = []   # Real part of complex number
    ai = []   # Imaginary part of complex number
    for line in gm.split('\n'):
        parse = re.search(r'\s*(\S+)\s*(\S+)\s*(\S+)', line)
        if parse:
            [ ix, real, imag ] = [ parse.group(1), parse.group(2), parse.group(3) ]

            # E.g. if index ix=3, we should be about to push the fourth number in each array
            assert len(ar) == int(ix); ar = ar + [ real ]
            assert len(ai) == int(ix); ai = ai + [ imag ]

    if DBG:
        print('real', ar)
        print('imag', ai)
        print("     _________GOLDEN MODEL_______")
        for ix in range(0,len(ar)):
            print(f"    {ix:4d} {ar[ix]:>9s}    {ai[ix]:>9s}")

    return (ar + ai)


def print_results(rtype, npoints, gm_results, fft_results, DBG=0):
    '''Compare GM/FFT real or imaginary results according to rtype "Real" or not'''

    nerrors   = 0;
    nwarnings = 0;

    # E.g. if npoints=8, then [0..7] are real nums and [8,15] are corresponding imag pts.
    npoints=int(npoints)
    [ first, last ] = [0, npoints] if (rtype == "Real") else [npoints, 2*npoints]

    print(f"{rtype} results:      GM          FFT")

    # Each result MUST match to at least one decimal place;
    # emit warning if errors at two or three d.p.

    for i in range(first, last):

        # Format with three, two and one sig figs
        g = gm_results[i]; f = fft_results[i]

        # Enable this block to test the warnings and errors
        if (0):
            if (i == 11): f = -0.415  # 3-fig warn vs. -2.414
            if (i == 13): f =  0.424  # 2-fig warn vs.  0.414
            if (i == 15): f =  2.511  # ERROR vs. 2.414

        [gr3,fr3] = sigfig(g, f, '%.3f', '[-]0[.]000','0.000')
        [gr2,fr2] = sigfig(g, f, '%.2f', '[-]0[.]00' ,'0.00' )
        [gr1,fr1] = sigfig(g, f, '%.1f', '[-]0[.]0'  ,'0.0'  )

        # ERROR if no match @ one sig figs.
        # Warning if no match @ two or three sig figs.

        status = ""
        if DBG: print(i, gr1, fr1)

        if (gr1 != fr1):
                status = f" *** ERROR '{gr1}' ne '{fr1}'";
                nerrors = nerrors + 1

        elif (gr2 != fr2):
            status = " *** WARNING mismatch at 2nd sig fig";
            nwarnings = nwarnings + 1

        elif (gr3 != fr3):
            status = " *** WARNING mismatch at 3rd sig fig";
            nwarnings = nwarnings + 1

        print(f"             {gr3:>9s}    {fr3:>9s}{status}")

    print('')
    return nerrors

def sigfig(g, f, fmt, neg, pos, DBG=0):
    '''
    E.g. if (g,f)=(0.1234,5.6789) and fmt="%.1", return ("0.1","5.6")
    Also, disallow minus-zero, i.e. turn e.g. "-0.00" into "0.00"
    '''
    gr = fmt % float(g); fr = fmt % float(f)

    # Oh this is annoying; must canonicalize e.g. "-0.000" => "0.000"
    # fr='-0.0'; fr='-0.00'; fr='-0.000'
    fr = re.sub(neg, pos, fr); gr = re.sub(neg, pos, gr)
    
    # Debugging
    if DBG: print(f'GM  {fmt} = {gr} ({g})')
    if DBG: print(f'FFT {fmt} = {fr} ({f})')
    if DBG: print('')

    # return [gr.strip(), fr.strip()] # Don't need the strip anymore, right?
    return [gr, fr]

# Same as shell `backtick` e.g. result=`echo foo | sed 's/oo/ox/'`
#   shell: result=`echo foo | sed 's/oo/ox/'`
#   this:  result= my_syscall("echo foo | sed 's/oo/ox/'")
def my_syscall(cmd):
    '''Example: [status,stdout,stderr]=my_syscall("echo foo | sed 's/oo/ox/'")
    '''
    from subprocess import run, PIPE
    p = run(cmd, shell=True, stdout=PIPE, stderr=PIPE)
    #
    [exit_status, stdout, stderr] = [
        p.returncode, p.stdout.decode(), p.stderr.decode() ]
    #
    return stdout

main()


